### ISSUE TITLE
Failing to make plv8-2.3.3 for PostgreSQL 11

### ISSUE THREAD

**Issue by Fanatique1337** [plv8/plv8js#371]
**Issue**
I'm getting a few errors everytime I run
```
make -f Makefile.shared
```
thus not allowing me to compile plv8 for Debian 10 / PostgreSQL 11.

**Setup**
Debian 10
PostgreSQL 11.5 (Debian 11.5-1+deb10u1)

**Detail**
After building V8 (using depot_tools) on Debian 10, I try to make plv8-2.3.3 for pg11. 

The following fails for me:
```
wget https://github.com/plv8/plv8/archive/v2.3.3.tar.gz
tar -xzf v2.3.3.tar.gz
cd plv8-2.3.3
make -f Makefile.shared
```
And I get the following errors:
```
g++ -Wall -O2 -std=c++11 -fno-rtti  -I. -I./ -I/usr/include/postgresql/11/server -I/usr/include/postgresql/internal  -Wdate-time -D_FORTIFY_SOURCE=2 -D_GNU_SOURCE -I/usr/include/libxml2  -I/usr/include/mit-krb5 -fPIC -c -o plv8.o plv8.cc
plv8.cc: In member function ‘void Converter::Init()’:
plv8.cc:1699:26: error: base operand of ‘->’ has non-pointer type ‘FormData_pg_attribute’
   if (m_tupdesc->attrs[c]->attisdropped)
                          ^~
In file included from /usr/include/postgresql/11/server/postgres.h:46,
                 from plv8.h:19,
                 from plv8.cc:8:
plv8.cc:1702:55: error: base operand of ‘->’ has non-pointer type ‘FormData_pg_attribute’
   m_colnames[c] = ToString(NameStr(m_tupdesc->attrs[c]->attname));
                                                       ^~
/usr/include/postgresql/11/server/c.h:576:25: note: in definition of macro ‘NameStr’
 #define NameStr(name) ((name).data)
                         ^~~~
plv8.cc:1714:29: error: base operand of ‘->’ has non-pointer type ‘FormData_pg_attribute’
          m_tupdesc->attrs[c]->atttypid,
                             ^~
plv8.cc: In member function ‘v8::Local<v8::Object> Converter::ToValue(HeapTuple)’:
plv8.cc:1737:26: error: base operand of ‘->’ has non-pointer type ‘FormData_pg_attribute’
   if (m_tupdesc->attrs[c]->attisdropped)
                          ^~
plv8.cc: In member function ‘Datum Converter::ToDatum(v8::Handle<v8::Value>, Tuplestorestate*)’:
plv8.cc:1785:27: error: base operand of ‘->’ has non-pointer type ‘FormData_pg_attribute’
    if (m_tupdesc->attrs[c]->attisdropped)
                           ^~
plv8.cc:1807:26: error: base operand of ‘->’ has non-pointer type ‘FormData_pg_attribute’
   if (m_tupdesc->attrs[c]->attisdropped)
                          ^~
make: *** [Makefile.shared:87: plv8.o] Error 1
```

Should I try installing a different version of plv8? Or is the problem somewhere else?

**Comment by JerrySievert** [comment]
first off, you should not be building plv8 using `make -f Makefile.shared`.  v8 is not designed to compile and link against a shared v8, and thus you should use `make` instead which will download and build the proper version of v8 with the proper settings and configuration for plv8.

seconds, 2.3.3 is pretty old, even for the 2.3 branch, clocking in at over a year and a half old.  current version in the 2.3 branch is 2.3.13.  since 2.3.3, postgres has moved from 10.0 to 12.0, with breaking changes that require extensions to be changed to build and run in each major postgres version.  for the 2.3 branch and postgres 11, you'll find these plv8 changes as part of 2.3.7: https://github.com/plv8/plv8/blob/v2.3.13/Changes#L25-L27  

I'd recommend updating to 2.3.13, which provides the same features and functionality as 2.3.3 but updated to work with newer versions of postgres that introduce breaking api changes.

**Comment by milesrichardson** [comment]
Hi,

I was able to get 2.3.13 to build. You can take a look at my fork here: https://github.com/milesforks/plv8 (maybe branch https://github.com/plv8/plv8/compare/r3.0alpha...milesforks:r3.0alpha-link-icu for 3.0alpha, but there is also https://github.com/plv8/plv8/compare/r3.0alpha...milesforks:v2.3.13-link-icu which worked for me building 2.3.13). Also maybe poke around at the other branches and the dockerfiles there to see how I got the makefiles to work for me.

The only change I had to make was to support ICU, I had to add the missing ICU static libraries (I've been meaning to put this into a PR) to the Makefile, and copy the ICU data from the v8 build

I also have a WIP branch upgrading v8 to 7.7 but I've gotten distracted the last week: https://github.com/plv8/plv8/compare/r3.0alpha...milesforks:r3.0alpha-link-icu-bump-v8 I resolved all the compiler errors, got it to build and install, but tests still fail. I think I made a few logic errors with JSON, and I commented out some things that still need to be upgraded to the 7.7 API. (disclaimer: not usually a C++ programmer). @JerrySievert Feel free to use that branch as a starting point if you want to take over. I'm not sure I'll be able to get to it within the next few weeks (and not sure the changes are even done correctly)

**Comment by JerrySievert** [comment]
@milesrichardson I'm guessing you didn't know about https://github.com/plv8/plv8/tree/r3.0alpha/contrib ? :) compiling `make USE_ICU=1` should build v8 with ICU, which gets you ICU configured and built as part of v8 directly.

I'd definitely accept a 7.7 PR.

**Comment by Fanatique1337** [comment]
@JerrySievert My bad for not checking the changelogs first. PLV8 2.3.7 compiles very neatly for PostgreSQL 11 and it works fine for now.

**Comment by milesrichardson** [comment]
@JerrySievert `USE_ICU=1` builds v8 with ICU, yes. But there was nothing linking the actual ICU static libraries it produced to the plv8 build. Also, the ICU data files in contrib were for a different version of ICU than the data files that v8 builds and places in the third_party directory. I had to make sure to copy the latter. 

I should have some time next week. I'll try to at least get the ICU changes cleaned up in a PR (and demonstrate what I mean). The 7.7 upgrade might take a bit longer to finish, but I did at least get it to build which seemed to be the hard part.

Cheers
